lib/src/constants.dart

```diff
@@ -0,0 +1,171 @@
+  bool get boolValue;
+
+  /// Returns whether this constant represents an `int` literal.
+  ///
+  /// If `true`, [intValue] will return an `int` (not throw).
+  bool get isInt;
+
+  /// Returns this constant as an `int` value.
+  ///
+  /// Throws [FormatException] if [isInt] is `false`.
+  int get intValue;
+
+  /// Returns whether this constant represents a `String` literal.
+  ///
+  /// If `true`, [stringValue] will return a `String` (not throw).
+  bool get isString;
+
+  /// Returns this constant as an `String` value.
+  ///
+  /// Throws [FormatException] if [isString] is `false`.
+  String get stringValue;
+
+  /// Returns whether this constant represents `null`.
+  bool get isNull;
+
+  /// Reads[ field] from the constant as another constant value.
+  Constant read(String field);
+
+  /// Reads [field] from the constant as a boolean.
+  ///
+  /// If the resulting value is `null`, uses [defaultTo] if defined.
+  bool readBool(String field, {bool defaultTo()});
+
+  /// Reads [field] from the constant as an int.
+  ///
+  /// If the resulting value is `null`, uses [defaultTo] if defined.
+  int readInt(String field, {int defaultTo()});
+
+  /// Reads [field] from the constant as a string.
+  ///
+  /// If the resulting value is `null`, uses [defaultTo] if defined.
+  String readString(String field, {String defaultTo()});
+}
+
+/// Implements a [Constant] representing a `null` value.
+class _NullConstant implements Constant {
+  const _NullConstant();
+
+  @override
+  bool get boolValue => throw new FormatException('Not a bool', 'null');
+
+  @override
+  int get intValue => throw new FormatException('Not an int', 'null');
+
+  @override
+  String get stringValue => throw new FormatException('Not a String', 'null');
+
+  @override
+  bool get isBool => false;
+
+  @override
+  bool get isInt => false;
+
+  @override
+  bool get isNull => true;
+
+  @override
+  bool get isString => false;
+
+  @override
+  Constant read(_) => this;
+
+  @override
+  bool readBool(_, {bool defaultTo(): _alwaysNull}) => defaultTo();
+
+  @override
+  int readInt(_, {int defaultTo(): _alwaysNull}) => defaultTo();
+
+  @override
+  String readString(_, {String defaultTo(): _alwaysNull}) => defaultTo();
+}
+
+/// Default implementation of [Constant].
+class _Constant implements Constant {
+  final DartObject _object;
+
+  const _Constant(this._object);
+
+  @override
+  bool get boolValue => isBool
+      ? _object.toBoolValue()
+      : throw new FormatException('Not a bool', _object);
+
+  @override
+  int get intValue => isInt
+      ? _object.toIntValue()
+      : throw new FormatException('Not an int', _object);
+
+  @override
+  String get stringValue => isString
+      ? _object.toStringValue()
+      : throw new FormatException('Not a String', _object);
+
+  @override
+  bool get isBool => _object.toBoolValue() != null;
+
+  @override
+  bool get isInt => _object.toIntValue() != null;
+
+  @override
+  bool get isNull => false;
```

Comment:
I think it's impossible to get into a state where that is true, but no harm doing this.